google/cloud/storage/examples/storage_object_samples.cc

```diff
@@ -174,6 +174,32 @@ void WriteObject(google::cloud::storage::Client client, int& argc,
   (std::move(client), bucket_name, object_name, desired_line_count);
 }
 
+void UpdateObjectMetadata(google::cloud::storage::Client client, int& argc,
+                          char* argv[]) {
+  if (argc != 5) {
+    throw Usage{
+        "update-object-metadata <bucket-name> <object-name> <key> <value>"};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto object_name = ConsumeArg(argc, argv);
+  auto key = ConsumeArg(argc, argv);
+  auto value = ConsumeArg(argc, argv);
+  //! [update object metadata] [START storage_set_metadata]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string object_name,
+     std::string key, std::string value) {
+    gcs::ObjectMetadata meta =
+        client.GetObjectMetadata(bucket_name, object_name);
+    gcs::ObjectMetadata desired = meta;
+    desired.mutable_metadata().emplace(key, value);
```

Comment:
I kind of like the upsert method in metadata instead of chaining: `desired.upsert_metadata(key, value)`. I personally thought having `desidered[key].value`, which under the hood calls emplace on the map. 

My only comment for this PR.

Reply:
Created #1046 to implement those.